onmt/modules/copy_generator.py

```diff
@@ -259,3 +237,69 @@ def _compute_loss(self, batch, output, target, copy_attn, align,
             loss = loss.sum()
 
         return loss, stats
+
+
+class CopyGeneratorLossCompute(NMTLossCompute, CopyGeneratorLossComputeBase):
+    """Copy Generator Loss Computation."""
+    def __init__(self, criterion, generator, tgt_vocab, normalize_by_length,
+                 lambda_coverage=0.0):
+        super(CopyGeneratorLossCompute, self).__init__(
+            criterion, generator, lambda_coverage=lambda_coverage)
+        self.tgt_vocab = tgt_vocab
+        self.normalize_by_length = normalize_by_length
+
+    def _make_shard_state(self, batch, output, range_, attns):
+        """See base class for args description."""
+        if getattr(batch, "alignment", None) is None:
+            raise AssertionError("using -copy_attn you need to pass in "
+                                 "-dynamic_dict during preprocess stage.")
```

Comment:
There is no `-dynamic_dict` anymore since we've merged preprocess to training. `-copy_attn` should now be used in all places indicating `-dynamic_dict`.
This assertion seems indeed obsolete.

Reply:
This one needs to be addressed.

Reply:
I removed the assertion